npeps/npep-133-fqdn-egress-selector.md

```diff
@@ -0,0 +1,438 @@
+    particular workload.
+  * This selector provides no capability to detect traffic destined for
+    different domains backed by the same IP (e.g. CDN or load balancers).
+* This enhancement does not provide a mechanism for selecting in-cluster
+  endpoints using FQDNs. To select Pods, Nodes, or the API Server,
+  AdminNetworkPolicy has other more specific selectors.
+  * Using the FQDN selector to refer to other Kubernetes endpoints, while not
+    explicitly disallowed, is not defined by this spec and left up to individual
+    providers. Trying to allow traffic to the following domains is NOT
+    guaranteed to work:
+    * `my-svc.my-namespace.svc.cluster.local` (the generated DNS record for a
+      Service as defined
+      [here](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#services))
+    * `pod-ip-address.my-namespace.pod.cluster.local` (the generated DNS record
+      for a Pod as defined
+      [here](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pods))
+* This enhancement does not add any new mechanisms for specifying how traffic is
+  routed to a destination (egress gateways, alternative SNAT IPs, etc). It just
+  adds a new way of specifying packets to be allowed or dropped on the normal
+  egress data path.
+* This enhancement does not require any mechanism for securing DNS resolution
+  (e.g. DNSSEC or DNS-over-TLS). Unsecured DNS requests are expected to be
+  sufficient for looking up FQDNs.
+
+## Introduction
+
+FQDN-based egress controls are a common enterprise security practice.
+Administrators often prefer to write security policies using DNS names such as
+“www.kubernetes.io” instead of capturing all the IP addresses the DNS name might
+resolve to. Keeping up with changing IP addresses is a maintenance burden, and
+hampers the readability of the network policies.
+
+## User Stories
+
+* As a cluster admin, I want to allow all Pods in the cluster to send traffic to
+  an external service specified by a well-known domain name. For example, all
+  Pods must be able to talk to `my-service.com`.
+
+* As a cluster admin, I want to allow Pods in the "monitoring" namespace to be
+  able to send traffic to a logs-sink, hosted at `logs-storage.com`
+
+* As a cluster admin, I want to allow all Pods in the cluster to send traffic to
+  any of the managed services provided by my Cloud Provider. Since the cloud
+  provider has a well known parent domain, I want to allow Pods to send traffic
+  to all sub-domains using a wild-card selector -- `*.my-cloud-provider.com`
+
+### Future User Stories
+
+These are some user stories we want to keep in mind, but due to limitations of
+the existing Network Policy API, cannot be implemented currently. The design
+goal in this case is to ensure we do not make these unimplementable down the
+line.
+
+* As a cluster admin, I want to block all cluster egress traffic by default, and
```

Comment:
This is possible now that the egress npep has merged right, or am I mis-understanding?

Reply:
I think the idea we wanted to capture is that we need FQDN Selectors at the namespaced NetworkPolicy layer. This was a suggestion from @danwinship to basically remind ourselves that some sort of FQDN selector will eventually end up in NetPolV2 and we don't want to select any behavior here which can't be implemented in a future implementation.

I'm not sure this user-story does a great job of conveying that... I'll try to think of a better phrasing, but suggestions very much welcome.

Reply:
Perhaps our user story (which really is a variant of the first user story) would clear anything up?

As a cluster admin, I want to allow certain pods in a namespace to access
an external service specified by a well-known domain name. The service is hosted on a cloud and changes IP-addresses often.

Reply:
@smuda I agree, I think this is well captured by the first user story

@astoycos PTAL, I re-worded a little bit, to  hopefully make it easier to understand.